lib/http_aws_sigv4.c

```diff
@@ -66,6 +66,89 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+/*
+ * Decide in an encoding-independent manner whether a character in an
+ * URL must be escaped. The same criterion must be used in strlen_uri()
+ * and strcpy_uri().
+ */
+static bool urlchar_needs_escaping(int c, bool in_query)
+{
+  return !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
+           (c >= '0' && c <= '9') || c == '-' || c == '.' || c == '_'
+           || c == '~' ||
+           (!in_query && c == '/') ||
+           (in_query && c == '&'));
+}
+
+/*
+ * strlen_uri() returns the length of the given URI
+ * this fucntion is a modified version of strlen_url in urlapi.c
+ */
+static size_t strlen_uri(const char *url, bool in_query)
+{
+  const unsigned char *ptr;
+  size_t newlen = 0;
+  int count_equal = 0;
+
+  for(ptr = (unsigned char *)url; *ptr; ptr++) {
+    switch (*ptr) {
+    case '=':
+      if(!in_query || count_equal++)
+        newlen += 2;
+      newlen++;
+      break;
+    case '&':
+      count_equal = 0;
+      /* FALLTHROUGH */
+    default:
+      if(urlchar_needs_escaping(*ptr, in_query))
+        newlen += 2;
+      newlen++;
+    }
+  }
+  return newlen;
+}
+
+/* strcpy_uri() copies a url to a output buffer and URL-encodes
+ * the URI acordingly
+ * this fucntion is a modified version of strcpy_url in urlapi.c
+ */
+static void strcpy_uri(char *optr, const char *url, bool in_query)
+{
+  /* we must add this with whitespace-replacing */
+  const unsigned char *iptr;
+  int count_equal = 0;
+
+  for(iptr = (unsigned char *)url;    /* read from here */
+      *iptr;         /* until zero byte */
+      iptr++) {
+
+    switch(*iptr) {
+    case '=':
+      if(!in_query || count_equal++) {
+        msnprintf(optr, 4, "%%%02X", *iptr);
+        optr += 3;
+      }
+      else
+        *optr++=*iptr;
+      break;
+    case '&':
+      count_equal = 0;
```

Comment:
I don't understand why it is curl's job to encode the equals signs like this. The user has provided a URL to use. Why would the user provide a wrong one for curl to re-encode like this?

Reply:
https://docs.aws.amazon.com/AmazonS3/latest/API/images/sigV4-using-query-params.png
In this document AWS say that you need to encode every parameter and arguments of the query string. problem is that curl seems to contain the whole query string as a single, so I had to find a way to determine which part need to be encode, and which part of need to be left as it.

Reply:
@bagder, AWS doesn't recommend using standard UriEncode functions provided by development platform may not work because of differences in implementation and related ambiguity in the underlying RFCs.

Reply:
2 small details:
the equal is if typicalle an user have something like `https://api.region.provider/Call&Argument====&Argument2=2=` in which case, if I'm correct, all equal but the first after `Argument` and `Argument` must be encode.

Also I could patch the functions in url API, but it would add a lot of if/else which might be ugly (for examples curl use lovercase for hexa and AWS require upper case), also as @drTr0jan  said, AWS advise to reimplement the encoding.

Reply:
That's borderline crazy since a URL is already URL encoded as otherwise it's not a URL. Does the spec also clearly say what bytes it needs to re-encode?

Reply:
BTW, doesn't that also imply that you should first URL *decode* the query string so that you get it truly canonicalized? Otherwise existing URL-encoded characters like %41 might thwart this.

Reply:
humm I think the issue this PR is trying to fix is, when curl is call with "https://${ES_DOMAIN_ENDPOIN}/tempo_data-*/event/_mapping" instead of "https://${ES_DOMAIN_ENDPOIN}/tempo_data-%2A/event/_mapping"
So I guess it might not be curl job to encode those URL.
It might still be an useful feature, maybe add an option to tell curl that the URL is not encoded yet ?

Reply:
Now I think about it, I do think the AWS doc is about not yet encoded URL, (so not URL).
also the issue is about using "https://${ES_DOMAIN_ENDPOIN}/tempo_data-*/event/_mapping" instead of "https://${ES_DOMAIN_ENDPOIN}/tempo_data-%2A/event/_mapping".

Reply:
though it might still be useful to add an option to force encoding ? (in which case we won't have to deal with already encoded URL)

Reply:
I think the issue here is that they want make the url canonical, and to do that you need to first URL decode and the URL encode because curl always operate on a URL as input and that means it is encoded. So a `*` could be passed in as a `*` or as `%2a` or as `%2A`. They're all fine according to the URL syntax. But they will then genrate different hashes and that's not what a user wants.